### ISSUE TITLE
404 page is cached

### ISSUE THREAD

**Issue by fritzmg** [contao/contao#231]
**Affected version(s)**
4.4.30 (I have not tested 4.6.10 yet).

**Description** 
If you have defined a `404` page in your site structure, then every request that results in a 404 response (i.e. every request that will show this page) will be put into the `http_cache`. 

This can blow up the HTTP cache due to bots accessing old non-existent URLs, bot attacks trying to access known vulnerabilities of other CMS etc. [One user on the forum](https://community.contao.org/de/showthread.php?72908-Contao-Core-4-4-nachtr%C3%A4glich-keine-Installation-von-default-Modulen-m%C3%B6glich) reported that his HTTP cache was 30 GiB in size.

**How to reproduce**  

1. Enable the page cache in your website root.
2. Verify that a request which results in a 404 response will not be put into the cache.
3. Create a 404 page in your site structure (and do not override the cache settings).
4. Create another request that results in a 404 response.

The 404 page will be present in the HTTP cache.

Imho it should not be necessary to manually override the cache settings on your 404 or 403 pages. There is no point in ever caching 404 or 403 response for each and every URL, is there?

**Comment by Paddy0174** [comment]
+1 for never caching 403 and 404 pages

**Comment by leofeyer** [comment]
Hm, the 40* pages have the fields to define a cache timeout, so I guess this is a configuration error in the first place.

@contao/developers Any objections against removing the fields and never caching those pages?

**Comment by bytehead** [comment]
+1 for removing and never caching.

**Comment by Toflar** [comment]
From a caching point of view, it's perfectly fine to cache 404 and 403 pages. The behaviour is correct.
But I see the issue with bloating the reverse proxy cache. We might want to force `Cache-Control: private` so it's cached on the client for subsequent identical requests but not in the reverse proxy?

**Comment by aschempp** [comment]
Well the user can configure if a 404 should be cached on the server, right? So the correct configuration would be to only allow client cache configuration for 404?

**Comment by Toflar** [comment]
Yes, remove it from these pages but at the same time also make sure existing values are replaced. Otherwise an update would not solve the issue because in the DB there's still the old configuration.

**Comment by aschempp** [comment]
Not necessarily remove it, but just overwrite the value in the page type.

**Comment by leofeyer** [comment]
I wonder if our fix is correct, as [Varnish caches 404 pages as well](https://www.section.io/blog/varnish-cache-404-errors/#what-should-happen-when-varnish-cache-sees-a-404-error) unless explicitly told otherwise. If we removed the "server cache timeout" dropdown and always set the cache timeout value to `0`, we would take away the possibility to cache error pages if someone wants to.

**Comment by Toflar** [comment]
None of our users will ever get this right. In Contao, it makes no sense to do so.

**Comment by leofeyer** [comment]
We do have some pretty advanced users as well. 😄

**Comment by contaoacademy** [comment]
I think the problem is that normal users make the settings on the root page and it is simply inherited.

Possibly it could be programmed, that the 404-page is excluded from inheritance and the settings can only be set explicitly.

**Comment by leofeyer** [comment]
As discussed in Mumble on Januar 17th, we want the error pages to no longer inherit the cache timeouts from their parent pages as proposed by @contaoacademy.

**Comment by leofeyer** [comment]
So basically if the page does not have the `includeCache` flag enabled, we have to reset the cache timeouts:

```php
public function getResponse($objRootPage=null)
{
	/** @var PageModel $objPage */
	global $objPage;

	$obj401 = $this->prepare($objRootPage);
	$objPage = $obj401->loadDetails();

	// Reset inherited cache timeouts (see #231)
	if (!$objPage->includeCache)
	{
		$objPage->cache = 0;
		$objPage->clientCache = 0;
	}

	/** @var PageRegular $objHandler */
	$objHandler = new $GLOBALS['TL_PTY']['regular']();

	return $objHandler->getResponse($objPage)->setStatusCode(401);
}
```

@contao/developers Why is this issue assigned to the 4.8.0 milestone? Isn't this a bugfix?

> This can blow up the HTTP cache due to bots accessing old non-existent URLs, bot attacks trying to access known vulnerabilities of other CMS etc. One user on the forum reported that his HTTP cache was 30 GiB in size.

**Comment by aschempp** [comment]
is `$objPage->includeCache` not inherited from parent pages?

**Comment by leofeyer** [comment]
Nope.

**Comment by aschempp** [comment]
Sounds legit. We must also prevent the model from being stored in the DB now though.

**Comment by leofeyer** [comment]
Fixed in 888743ea7b002e8792155b9b9b68bb1f2397dfef.

> Sounds legit. We must also prevent the model from being stored in the DB now though.

The model instance is already detached because of `loadDetails()`.
